### PR TITLE
doc: don't use typehints for signatures.

### PR DESCRIPTION
**Pull request by jakevdp**
The reason to avoid this is that overloaded signatures can be very confusing.

This is the current documentation page for `jax.numpy.where`:
screenshot-2

This is the same page built with this PR:
screenshot-1

The downside is that for non-overloaded functions, this change causes type annotations to not be inserted into parameter descriptions. For example this is the current documentation page for `jax.scipy.linalg.inv`:
screenshot-4

And this is the same page built with this PR:
screenshot-3

I think the benefit in the case of `jnp.where` and other overloaded functions far outweighs the loss of annotations in other cases.

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -294,7 +294,11 @@
 # -- Extension configuration -------------------------------------------------
 
 # Tell sphinx autodoc how to render type aliases.
-autodoc_typehints = "description"
+
+# type aliases are not respected in overloaded function signatures, which leads to
```
Comment (superbobry) [inline docs/conf.py:298]: Is there an autodoc bug we can link to here?

Comment (jakevdp) [reply docs/conf.py:298]: No, i don’t think it’s a bug. The type signatures are displaying correctly, they’re just not helpful in most cases because of our long aliases.

Comment (superbobry) [reply docs/conf.py:298]: I think autodoc does not expand aliases in modules with `from __future__ import annotations`. 

We did this hack in Haiku because of that: https://github.com/google-deepmind/dm-haiku/blob/a7b7e73dae840153ecd828e97a64b6a875b168f7/docs/conf.py#L36, but since JAX is 3.9+ we can also just add the `__future__` import everywhere. Wdyt?

Comment (jakevdp) [reply docs/conf.py:298]: I saw that in the docs. We already have the future import in the relevant modules, it didn't seem to make a difference.

Comment (jakevdp) [reply docs/conf.py:298]: The aliasing works correctly in the `description`, but seems not to work in overloads.

We previously had `autodoc_typehints='description'`, but according to the [documentation](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_typehints), for overloaded functions this is the same as `autodoc_typehints='signature'`.

I think our only recourse to avoid the busy expansions in the `jnp.where` screenshot above is to set `autodoc_typehints='none'`.

Comment (superbobry) [reply docs/conf.py:298]: I've sent #22208 which seems to work locally.
